src/config/prettier-config-explorer/resolver.js

```diff
@@ -0,0 +1,132 @@
+}
+
+/**
+ * Searches for a configuration file in the specified directory.
+ * 
+ * @param {string} directory - The directory to search in.
+ * @returns {Promise<string|undefined>} - The path of the configuration file if found, otherwise undefined.
+ */
+async function searchConfigInDirectory(directory) {
+  for (const fileName of CONFIG_FILE_NAMES) {
+    const file = path.join(directory, fileName);
+
+    if (!(await fileExists(file))) {
+      continue;
+    }
+
+    if (
+      fileName !== "package.json" ||
+      (await isPackageJsonFileWithPrettierConfig(file))
+    ) {
+      return file;
+    }
+  }
+}
+
+class Resolver {
+  /** @type {Map<string, string>} */
+  #searchCache = new Map();
+
+  /** @type {Map<string, Promise<any>>} */
+  #loadCache = new Map();
+
+  /** @type {(configFile: string) => Promise<any>} */
+  #loader;
+
+  constructor({ loader }) {
+    this.#loader = loader;
+  }
+
+  /**
+   * Searches for a configuration file starting from a specified directory.
+   * 
+   * @param {string} startDirectory - The directory to start the search from.
+   * @param {{cache?: boolean, stopDirectory?: string}} [options={}] - The options to use.
+   * @returns {Promise<string|undefined>} - A promise that resolves to the path of the found configuration file, or null if not found.
+   */
+  async search(startDirectory, options = {}) {
+    const { cache, stopDirectory } = options;
+
+    let configFile;
+    const visitedDirectories = new Set();
+    
+    for (const directory of getParentDirectories(
+      startDirectory,
+      stopDirectory,
+    )) {
+      // If cache is enabled and the directory is cached, return the cached result.
+      if (cache && this.#searchCache.has(directory)) {
+        configFile = this.#searchCache.get(directory);
+        break;
+      }
+
+      visitedDirectories.add(directory);
+
+      configFile = await searchConfigInDirectory(directory);
+      if (configFile) {
+        break;
+      }
+    }
+
+    // If cache is enabled, cache the result for all visited directories.
+    if (cache) {
+      for (const directory of visitedDirectories) {
+        this.#searchCache.set(directory, configFile);
+      }
+    }
+
+    return configFile;
+  }
+
+  /**
+   * Loads the specified config file and returns a promise that resolves to the loaded file.
+   * 
+   * @param {string} configFile - The path to the config file.
+   * @param {{ cache?: boolean }} [options={}] - The options object.
+   * @returns {Promise<any>} A promise that resolves to the loaded file.
+   */
+  load(configFile, options = {}) {
+    const { cache } = options;
+
+    // If cache is enabled and the file is cached, return the cached result.
+    if (cache && this.#loadCache.has(configFile)) {
+      return this.#loadCache.get(configFile);
+    }
+
+    // Otherwise, load the file and cache the result if cache is enabled.
+    const promise = this.#loader(configFile);
+    if (cache) {
+      this.#loadCache.set(configFile, promise);
```

Comment:
This is incorrect, the cache must take stopDirectory into count, otherwise it will hit wrong cache,

eg we have config in  `/a` and first search `/a/b/c` with stopDirectoy set to `/a/b` then we will cache result for both `/a/b/c` and `/a/b` as undefined. Then we search `/a/b/c` with stopDirectoy set to `/` the result should find the file in `/a` not `undefined`.

The implementation in my branch also not correct, I can't use `undefined` since different file can have different path root on windows.

Reply:
This is a very good point. I will be adding tests to validate this.

Reply:
After a better thinking of this,

> The implementation in my branch also not correct, I can't use undefined since different file can have different path root on windows.


Turns out my implementation is corrert (but just happened correct, I didn't thought about differnt path root at first), since when the stopDirectory on Windows can be different dirver letter (`D:\\`), but they can't hit the wrong cache, since `D:\\` in in the cache key.

Reply:
I addressed this issue and added unit tests to validate the behavior. 
@fisker Let me know what you think.